[my-base] 2020-07-11 (1) - <wjtliso>

wjtliso opened the pull request:
在开这个 PR 之前请确认以下几点：

# rebase master 了吗？
`git rebase master`
- [x] 已检查

# PR 标题格式正确吗？（如果是提交作业的话）
`[<package name>] <deadline> (<number of submission>) - <username>`
比如说 `[my-base] Jun 20, 2020 (2) - carassius1014`
- [x] 已检查

wujietao-wamazing committed:
code move

wujietao-wamazing committed:
code move

wujietao-wamazing committed:
finish maybe

wujietao-wamazing committed:
finish maybe

wujietao-wamazing committed:
finish all instance

wujietao-wamazing committed:
finish all instance

wujietao-wamazing committed:
fix comment

wujietao-wamazing committed:
fix comment

wujietao-wamazing committed:
homework2

wujietao-wamazing committed:
homework2

wujietao-wamazing committed:
homework list

wujietao-wamazing committed:
homework list

wujietao-wamazing committed:
finish sortBy

wujietao-wamazing committed:
finish sortBy

wujietao-wamazing committed:
setting

wujietao-wamazing committed:
setting

wujietao-wamazing committed:
finish review the code

wujietao-wamazing committed:
finish review the code

wujietao-wamazing committed:
fix mconcat

wujietao-wamazing committed:
fix mconcat

carassius1014 committed:
Remove filterM_

carassius1014 committed:
Merge pull request #11 from momoyama-south-gate/develop

Remove filterM_

carassius1014 committed:
Export Alternative property tests

carassius1014 committed:
Merge pull request #12 from momoyama-south-gate/develop

Export Alternative property tests

wujietao-wamazing committed:
temp save

wujietao-wamazing committed:
homework2

wujietao-wamazing committed:
Merge branch 'wjtliso-report-box' of https://github.com/momoyama-south-gate/haskell-codespace into wjtliso-report-box

carassius1014 commented on my-base/src/My/Control/Applicative.hs:
```diff
@@ -197,4 +201,10 @@ some = undefined
 many :: Alternative f => f a -> f [a]
 many = undefined
```
TODO

carassius1014 commented on my-base/src/My/Data/Tuple.hs:
```diff
@@ -52,10 +52,16 @@ instance Functor ((,) a) where
 -- rhs = (0, \f -> f 0) <*> (1, const 0) = (1, 0)
 
-instance Monoid a => Monad ((,) a)
+instance Monoid a => Monad ((,) a) where
```
这个符合 Monad Law 么？

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -55,62 +57,102 @@ prop_Monad_Right_Id ::
 
-replicateM_ :: Applicative f => Int -> f a -> f ()
-replicateM_ = undefined
+--replicateM_ :: Applicative f => Int -> f a -> f ()
+--replicateM_ = undefined
```
TODO

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -55,62 +57,102 @@ prop_Monad_Right_Id ::
 
-zipWithM_ :: Applicative f => (a -> b -> f c) -> [a] -> [b] -> f ()
-zipWithM_ = undefined
+--zipWithM_ :: Applicative f => (a -> b -> f c) -> [a] -> [b] -> f ()
+--zipWithM_ = undefined
```
TODO

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -55,62 +57,102 @@ prop_Monad_Right_Id ::
 
 replicateM :: Applicative f => Int -> f a -> f [a]
-replicateM = undefined
+replicateM n f = liftA (replicate n) f
+  where 
+    replicate :: Int -> a -> [a]
+    replicate n a =
+      if n == 0
+      then []
+      else a : replicate (n-1) a
```
这个函数应该是执行 `f a` n次之后把每次的结果收集起来。
你现在的实现是执行 `f a` 一次然后把结果重复n遍。

carassius1014 commented on my-base/src/My/Control/Applicative.hs:
```diff
@@ -197,4 +201,10 @@ some = undefined
+-- prop> prop_Alternative_Left_Id @ZipList
+-- prop> prop_Alternative_Right_Id @ZipList
+-- prop> prop_Alternative_Assoc @ZipList
+instance Alternative ZipList where
+  empty = ZipList []
+  (<|>) (ZipList la) (ZipList lb) = ZipList (la P.++ lb)
```
`ZipList` 的 `Alternative` 应该是左边不为空的话即左边，否则取右边。
现在这个实现和 `Alternative []` 没有区别。

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -55,62 +57,102 @@ prop_Monad_Right_Id ::
 
 (=<<) :: Monad m => (a -> m b) -> m a -> m b
-(=<<) = undefined
+(=<<) = flip $ (>>=)
```
`$` 多余了。

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -55,62 +57,102 @@ prop_Monad_Right_Id ::
 
 (<=<) :: Monad m => (b -> m c) -> (a -> m b) -> (a -> m c)
-(<=<) = undefined
+(<=<) = flip $ (>=>)
```
`$` 多余了。

carassius1014 reviewed (changes requested):
先改这么多吧。

wujietao-wamazing committed:
1.fix instance Monoid a => Monad ((,) a)
2.fix replicateM
3.fix replicateM_
4.fix zipWithM_
5.fix instance Alternative ZipList
6.fix (=<<)
7.fix (<=<)

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -125,20 +125,15 @@ zipWithM f la lb = foldl (\fl (a,b) -> liftA2 (\y l -> y : l) (f a b) fl) (pure
---zipWithM_ :: Applicative f => (a -> b -> f c) -> [a] -> [b] -> f ()
---zipWithM_ = undefined
+zipWithM_ :: Applicative f => (a -> b -> f c) -> [a] -> [b] -> f ()
+zipWithM_ f la lb  = const (pure ()) (zipWithM f la lb)
```
用 `void`

carassius1014 commented on my-base/src/My/Control/Monad.hs:
```diff
@@ -125,20 +125,15 @@ zipWithM f la lb = foldl (\fl (a,b) -> liftA2 (\y l -> y : l) (f a b) fl) (pure
+replicateM n f = liftA2 (\a l -> a : l) f (replicateM (n-1) f)
+
+replicateM_ :: Applicative f => Int -> f a -> f ()
+replicateM_ n f = const (pure ()) (replicateM_ n f)
```
用 `void`

carassius1014 reviewed (changes requested):
除了该用 `void` 的地方没问题。

wujietao-wamazing committed:
1. fix void
2. fix some
3. fix many
4. fix Alternative Maybe

carassius1014 reviewed (approved):
看起来 OK
